brainframe_qt/api_utils/connection_manager.py

```diff
@@ -0,0 +1,154 @@
+    _CONNECT_TIMEOUT = 5  # s
+
+    connection_state_changed = pyqtSignal(object)
+    license_state_changed = pyqtSignal(bf_codecs.LicenseInfo.State)
+
+    connection_error = pyqtSignal(Exception)
+
+    class ConnectionState(Enum):
+        UNCONFIGURED = auto()
+        """Credentials not accessed, or in need of update"""
+        UNCONNECTED = auto()
+        """No communication with server yet"""
+        LICENSE_UNVALIDATED = auto()
+        """License not yet validated"""
+        LICENSE_EXPIRED = auto()
+        """Current license is expired"""
+        LICENSE_INVALID = auto()
+        """Current license invalid"""
+        LICENSE_MISSING = auto()
+        """Current license is missing"""
+        CONNECTED = auto()
+        """Connected to server and ready to use API"""
+        RUNTIME_ERROR = auto()
+
+    def __init__(self, *, parent: QObject):
+        super().__init__(parent=parent)
+
+        self._connection_state = self.ConnectionState.UNCONFIGURED
+        self._connection_configuration: Optional[ConnectionConfiguration] = None
+
+        self.server_settings = ServerSettings()
+
+        self._init_signals()
+
+    def run(self) -> None:
+        while not self.isInterruptionRequested():
+            if self.connection_state is self.ConnectionState.CONNECTED:
+                QThread.msleep(self._CONNECTED_SLEEP_TIME)
+            elif self.connection_state is self.ConnectionState.UNCONFIGURED:
+                self._get_configuration()
+            elif self.connection_state is self.ConnectionState.UNCONNECTED:
+                self._communicate_with_server()
+            elif self.connection_state in [
+                self.ConnectionState.LICENSE_UNVALIDATED,
+                self.ConnectionState.LICENSE_EXPIRED,
+                self.ConnectionState.LICENSE_INVALID,
+                self.ConnectionState.LICENSE_MISSING
+            ]:
+                self._validate_license()
+            elif self.connection_state is self.ConnectionState.RUNTIME_ERROR:
+                QThread.msleep(self._CONNECTED_SLEEP_TIME)
+            else:
+                exc = RuntimeError(f"Unknown ConnectionState {self.connection_state}")
+                self._handle_error(exc)
+
+    def _init_signals(self) -> None:
+        self.server_settings.value_changed.connect(self._handle_settings_change)
+
+    @property
+    def connection_state(self) -> ConnectionState:
+        return self._connection_state
+
+    @connection_state.setter
+    def connection_state(self, connection_state: ConnectionState) -> None:
+        self._connection_state = connection_state
+        self.connection_state_changed.emit(self._connection_state)
+
+    def invalidate_config(self) -> None:
+        """Force the ConnectionManager to restart the authentication process"""
+        self.connection_state = self.ConnectionState.UNCONFIGURED
+
+    def _get_configuration(self) -> None:
+        url = self.server_settings.server_url
```

Comment:
Right now on the UI if you uncheck the "Server uses configuration", does it clear the server_username and server_password? I don't believe it does, and I think the client would still be trying to connect with that information, even though it appears to be unchecked in the UI

Reply:
It doesn't _clear_ that information client-side, but it also doesn't use it/store it.

https://github.com/aotuai/brainframe-qt/blob/9c8574b3220ad9b00bd9f6fa2ea39d61400f772b/brainframe_qt/ui/dialogs/server_configuration/server_configuration.py#L145-L150